tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1461 @@
+            .BuildRequest(
+              credentials_.GetAccessKey(),
+              credentials_.GetSecretKey(),
+              owner->aws_region_,
+              kEmptySha256,
+              now);
+
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_REQUESTS) << "WriteTask (Peek): " << request;
+
+    auto future = owner->transport_->IssueRequest(request, {});
+    future.ExecuteWhenReady(
+      [self = IntrusivePtr<WriteTask>(this)](
+          ReadyFuture<HttpResponse> response) {
+      self->OnPeekResponse(response.result());
+    });
+  }
+
+  void OnPeekResponse(const Result<HttpResponse>& response) {
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_RESPONSES && response.ok())
+        << "WriteTask (Peek) " << *response;
+
+    if(!response.ok()) {
+      promise.SetResult(response.status());
+      return;
+    }
+
+    TimestampedStorageGeneration r;
+    r.time = absl::Now();
+    switch (response.value().status_code) {
+      case 304:
+        // Not modified implies that the generation did not match.
+        [[fallthrough]];
+      case 412:
+        // Failed precondition implies the generation did not match.
+        r.generation = StorageGeneration::Unknown();
+        promise.SetResult(r);
+        return;
+      case 404:
+        if (!StorageGeneration::IsUnknown(options.if_equal) &&
+            !StorageGeneration::IsNoValue(options.if_equal)) {
+          r.generation = StorageGeneration::Unknown();
+          promise.SetResult(r);
+          return;
+        }
+      default:
+        break;
+    }
+
+    DoPut();
+  }
+
+  void DoPut() {
+    // TODO(sjperkins). Introduce S3 requester_pays logic here
+    // Assume that if the user_project field is set, that we want to provide
+    // it on the uri for a requestor pays bucket.
+    // AddUserProjectParam(&upload_url, true, owner->encoded_user_project());
+
+    // TODO(sjperkins).
+    // This was changed from POST to PUT as a basic POST does not work
+    // Some more headers need to be added to allow POST to work:
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-authentication-HTTPPOST.html
+    upload_url_ = tensorstore::StrCat(owner->endpoint_, "/", encoded_object_name);
+
+    start_time_ = absl::Now();
+    auto [content_sha256, checksum_sha256] = payload_sha256(value);
+    auto [content_md5, checksum_md5] = payload_md5(value);
+    expected_etag_ = checksum_md5;
+
+    auto request = S3RequestBuilder("PUT", upload_url_)
+            .AddHeader("Content-Type: application/octet-stream")
+            .AddHeader(tensorstore::StrCat("Content-Length: ", value.size()))
+            .AddHeader(absl::StrCat("host: ", owner->host_))
+            .AddHeader(absl::StrCat("x-amz-content-sha256: ", content_sha256))
+            .AddHeader(absl::StrCat("x-amz-checksum-sha256: ", checksum_sha256))
+            .AddHeader(absl::StrCat("x-amz-meta-tensorstore-id: ", content_sha256))
```

Comment:
`x-amz-checksum-sha256`  and `x-amz-meta-tensorstore-id` are probably both YAGNI, but I added both in to illustrate

1. Setting the object checksum
2. Setting custom object metadata. (https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html)

for further discussion